subprojects/docs/src/docs/userguide/buildEnvironment.adoc

```diff
@@ -13,135 +13,216 @@
+
+[NOTE]
+====
+If a project property is referenced but does not exist, an exception will be thrown and the build will fail.
+
+You should check for existence of optional project properties before you access them using the api:org.gradle.api.Project#hasProperty(java.lang.String)[] method.
+====
+
+[[sec:accessing_the_web_via_a_proxy]]
+=== Accessing the web through a HTTP proxy
+
+Configuring an HTTP or HTTPS proxy (for downloading dependencies, for example) is done via standard JVM system properties. These properties can be set directly in the build script; for example, setting the HTTP proxy host would be done with `System.setProperty('http.proxyHost', 'www.somehost.org')`. Alternatively, the properties can be <<sec:gradle_configuration_properties,specified in gradle.properties>>.
+
+.Configuring an HTTP proxy using `gradle.properties`
+====
+[source,properties]
+----
 systemProp.http.proxyHost=www.somehost.org
 systemProp.http.proxyPort=8080
 systemProp.http.proxyUser=userid
 systemProp.http.proxyPassword=password
 systemProp.http.nonProxyHosts=*.nonproxyrepos.com|localhost
-
 ----
-
-
 ====
 
 There are separate settings for HTTPS.
 
-.Configuring an HTTPS proxy
+.Configuring an HTTPS proxy using `gradle.properties`
 ====
-
-`gradle.properties`
-
 [source,properties]
 ----
-
 systemProp.https.proxyHost=www.somehost.org
 systemProp.https.proxyPort=8080
 systemProp.https.proxyUser=userid
 systemProp.https.proxyPassword=password
 systemProp.https.nonProxyHosts=*.nonproxyrepos.com|localhost
-
 ----
-
-
 ====
 
-We could not find a good overview for all possible proxy settings. One place to look are the constants in a file from the Ant project. Here's a https://git-wip-us.apache.org/repos/asf?p=ant.git;a=blob;f=src/main/org/apache/tools/ant/util/ProxySetup.java;hb=HEAD[link] to the repository. The other is a http://download.oracle.com/javase/7/docs/technotes/guides/net/properties.html[Networking Properties page] from the JDK docs. If anyone knows of a better overview, please let us know via the mailing list.
+You may need to set other properties to access other networks. Here are 2 references that may be helpful:
 
+* link:https://git-wip-us.apache.org/repos/asf?p=ant.git;a=blob;f=src/main/org/apache/tools/ant/util/ProxySetup.java;hb=HEAD[ProxySetup.java in the Ant codebase]
+* link:http://download.oracle.com/javase/7/docs/technotes/guides/net/properties.html[JDK 7 Networking Properties]
 
-[[sub:ntlm_authentication]]
 ==== NTLM Authentication
 
 If your proxy requires NTLM authentication, you may need to provide the authentication domain as well as the username and password. There are 2 ways that you can provide the domain for authenticating to a NTLM proxy:
 
 * Set the `http.proxyUser` system property to a value like `__domain__/__username__`.
 * Provide the authentication domain via the `http.auth.ntlm.domain` system property.
 
+[[sec:configuring_task_using_project_properties]]
+=== Configuring a task using project properties
```

Comment:
I think we should move this directly underneath "Project properties". It would give an example of the usage after explain what it is.

Reply:
Very good, will do

Reply:
done